### PR TITLE
Specs for object_lifetime_tracker_act

### PR DESCRIPTION
**Pull request by parth21999**

### CODE REVIEW

File: devdoc/object_lifetime_tracker.md [hunk 1]
```diff
@@ -241,3 +252,38 @@ MOCKABLE_FUNCTION(, void, object_lifetime_tracker_destroy_all_objects_for_key, O
...
+
+```c
+MOCKABLE_FUNCTION(, OBJECT_LIFETIME_TRACKER_ACT_RESULT, object_lifetime_tracker_act, OBJECT_LIFETIME_TRACKER_HANDLE, object_lifetime_tracker, const void*, key, void*, object, OBJECT_LIFETIME_TRACKER_ACTION_FUNCTION, action_function, void*, context);
+```
+
+`object_lifetime_tracker_act` performs an action on the object that was registered with the given key. The action is performed by calling the given `action_function` with the given `context`. The `action_function` is called under lock, so the user must ensure that it does not call any functions that may cause a deadlock.
+
+**OBJECT_LIFETIME_TRACKER_43_081: [** If `object_lifetime_tracker` is `NULL`, `object_lifetime_tracker_act` shall fail and return `OBJECT_LIFETIME_TRACKER_ACT_ERROR`. **]**
+
+**OBJECT_LIFETIME_TRACKER_43_082: [** If `key` is `NULL`, `object_lifetime_tracker_act` shall fail and return `OBJECT_LIFETIME_TRACKER_ACT_ERROR`. **]**
```
Comment (parth21999) [inline devdoc/object_lifetime_tracker.md:267]: >*OBJECT_ [](http://example.com/codeflow?start=1&length=8)

Removing SRS_ for now to suppress traceability.

File: tests/reals/real_object_lifetime_tracker.h [hunk 2]
```diff
@@ -31,6 +31,7 @@ void real_object_lifetime_tracker_destroy(OBJECT_LIFETIME_TRACKER_HANDLE object_
 OBJECT_LIFETIME_TRACKER_REGISTER_OBJECT_RESULT real_object_lifetime_tracker_register_object(OBJECT_LIFETIME_TRACKER_HANDLE object_lifetime_tracker, const void* key, void* object, DESTROY_OBJECT destroy_object, const void* destroy_context);
 OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT real_object_lifetime_tracker_unregister_object(OBJECT_LIFETIME_TRACKER_HANDLE object_lifetime_tracker, const void* key, void* object);
 void real_object_lifetime_tracker_destroy_all_objects_for_key(OBJECT_LIFETIME_TRACKER_HANDLE object_lifetime_tracker, const void* key);
+OBJECT_LIFETIME_TRACKER_ACT_RESULT real_object_lifetime_tracker_act(OBJECT_LIFETIME_TRACKER_HANDLE object_lifetime_tracker, const void* key, void* object, OBJECT_LIFETIME_TRACKER_ACTION_FUNCTION action_function, void* context);
```
Comment (anporumb) [inline tests/reals/real_object_lifetime_tracker.h:34]: >real_object_lifetime_tracker_act [](http://example.com/codeflow?start=35&length=32)

missing between lines 17 and 18

File: tests/reals/real_object_lifetime_tracker_renames.h [hunk 3]
```diff
@@ -5,7 +5,9 @@
 #define object_lifetime_tracker_register_object real_object_lifetime_tracker_register_object
 #define object_lifetime_tracker_unregister_object real_object_lifetime_tracker_unregister_object
 #define object_lifetime_tracker_destroy_all_objects_for_key real_object_lifetime_tracker_destroy_all_objects_for_key
+#define object_lifetime_tracker_act real_object_lifetime_tracker_act
 
 #define OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT real_OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT
 #define KEY_MATCH_FUNCTION_RESULT real_KEY_MATCH_FUNCTION_RESULT
 #define OBJECT_MATCH_FUNCTION_RESULT real_OBJECT_MATCH_FUNCTION_RESULT
+#define OBJECT_LIFETIME_TRACKER_ACT_RESULT real_OBJECT_LIFETIME_TRACKER_ACT_RESULT
```
Comment (anporumb) [inline tests/reals/real_object_lifetime_tracker_renames.h:13]: >real_OBJECT_LIFETIME_TRACKER_ACT_RESULT [](http://example.com/codeflow?start=43&length=39)

file must end with an empty line

File: inc/c_util/object_lifetime_tracker.h [hunk 4]
```diff
@@ -35,10 +35,19 @@ MU_DEFINE_ENUM(OBJECT_LIFETIME_TRACKER_REGISTER_OBJECT_RESULT, OBJECT_LIFETIME_T
 
 MU_DEFINE_ENUM(OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT, OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT_VALUES);
 
+#define OBJECT_LIFETIME_TRACKER_ACT_RESULT_VALUES \
+    OBJECT_LIFETIME_TRACKER_ACT_OK, \
+    OBJECT_LIFETIME_TRACKER_ACT_ERROR, \
+    OBJECT_LIFETIME_TRACKER_ACT_NOT_FOUND, \
+    OBJECT_LIFETIME_TRACKER_ACT_KEY_NOT_FOUND
```
Comment (anporumb) [inline inc/c_util/object_lifetime_tracker.h:42]: >OBJECT_LIFETIME_TRACKER_ACT_KEY_NOT_FOUND [](http://example.com/codeflow?start=4&length=41)

404 not found in specs. Is this (`OBJECT_LIFETIME_TRACKER_ACT_KEY_NOT_FOUND`)used?

**Comment by anporumb** [comment]
> MU_DEFINE_ENUM_STRINGS(OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT, OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT_VALUES)

Consider adding the strings for OBJECT_LIFETIME_TRACKER_ACT_RESULT too

---
Refers to: src/object_lifetime_tracker.c:61 in aabf2c0. [](commit_id = aabf2c0d089f7731f8ae707e56d85906d78ae828, deletion_comment = False)

File: devdoc/object_lifetime_tracker.md [hunk 5]
```diff
@@ -38,17 +38,28 @@ MU_DEFINE_ENUM(OBJECT_LIFETIME_TRACKER_REGISTER_OBJECT_RESULT, OBJECT_LIFETIME_T
 
 MU_DEFINE_ENUM(OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT, OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT_VALUES);
 
+#define OBJECT_LIFETIME_TRACKER_ACT_RESULT_VALUES \
+    OBJECT_LIFETIME_TRACKER_ACT_OK, \
+    OBJECT_LIFETIME_TRACKER_ACT_ERROR, \
+    OBJECT_LIFETIME_TRACKER_ACT_NOT_FOUND, \
+    OBJECT_LIFETIME_TRACKER_UNREGISTER_KEY_NOT_FOUND
```
Comment (anporumb) [inline devdoc/object_lifetime_tracker.md:45]: >OBJECT_LIFETIME_TRACKER_UNREGISTER_KEY_NOT_FOUND [](http://example.com/codeflow?start=4&length=48)

I find it weird to have `OBJECT_LIFETIME_TRACKER_UNREGISTER_KEY_NOT_FOUND` as an enum value in an enum called `OBJECT_LIFETIME_TRACKER_ACT_RESULT`. Is this intended?

**Comment by anporumb** [comment]
> #include <cstddef>

should be moved top of the file.

---
Refers to: tests/reals/real_object_lifetime_tracker.h:21 in aabf2c0. [](commit_id = aabf2c0d089f7731f8ae707e56d85906d78ae828, deletion_comment = False)

**Review by anporumb — DISMISSED** [review]
:shipit:

**Comment by parth21999** [comment]
> #include <cstddef>

Done

---
In reply to: [2023849472](https://github.com/Azure/c-util/pull/376#issuecomment-2023849472) [](http://example.com/codeflow?ancestors=2023849472)

---
Refers to: tests/reals/real_object_lifetime_tracker.h:21 in aabf2c0. [](commit_id = aabf2c0d089f7731f8ae707e56d85906d78ae828, deletion_comment = False)

File: tests/reals/real_object_lifetime_tracker.h (hunk 2 above)
Comment (parth21999) [reply tests/reals/real_object_lifetime_tracker.h:34]: Fixed.

File: tests/reals/real_object_lifetime_tracker_renames.h (hunk 3 above)
Comment (parth21999) [reply tests/reals/real_object_lifetime_tracker_renames.h:13]: Fixed.

File: inc/c_util/object_lifetime_tracker.h (hunk 4 above)
Comment (parth21999) [reply inc/c_util/object_lifetime_tracker.h:42]: Fixed.

File: devdoc/object_lifetime_tracker.md (hunk 5 above)
Comment (parth21999) [reply devdoc/object_lifetime_tracker.md:45]: Fixed.

**Comment by parth21999** [comment]
> MU_DEFINE_ENUM_STRINGS(OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT, OBJECT_LIFETIME_TRACKER_UNREGISTER_OBJECT_RESULT_VALUES)

Done

---
In reply to: [2023839884](https://github.com/Azure/c-util/pull/376#issuecomment-2023839884) [](http://example.com/codeflow?ancestors=2023839884)

---
Refers to: src/object_lifetime_tracker.c:61 in aabf2c0. [](commit_id = aabf2c0d089f7731f8ae707e56d85906d78ae828, deletion_comment = False)

**Review by BertKleewein — APPROVED** [review]
:shipit:

**Pull request merged**